uportal-war/src/main/java/org/apereo/portal/utils/cache/CacheHealthReporterService.java

```diff
@@ -0,0 +1,128 @@
+ * cache.  Those listeners quickly contact this service with the details.
+ * Periodically this service produces a report on bad things that are occurring,
+ * if any, and writes it to the log.
+ *
+ * @author drewwills
+ */
+@Service("cacheHealthReporterService")
+public class CacheHealthReporterService {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private enum Reports {
+        CACHE_SIZE_NOT_LARGE_ENOUGH {
+            @Override
+            protected void doReportInternal(Logger logger, List<ReportTuple> list) {
+                if (list.size() != 0) {
+                    final StringBuilder report = new StringBuilder();
+                    final Map<String,Integer> counts = new HashMap<>();
+                    for (ReportTuple tuple : list) {
+                        final String cacheName = tuple.getCache().getName();
+                        Integer newCount = 1;  // default
+                        Integer oldCount = counts.get(cacheName);
+                        if (oldCount != null) {
+                            newCount = ++oldCount;
+                        }
+                        counts.put(cacheName, newCount);
+                    }
+                    report.append("The following cache(s) have insufficient maxElementsInMemory;  " +
+                            "there must be room in the cache to hold every object of " +
+                            "the corresponding type in the portal:");
+                    for (Map.Entry<String,Integer> y : counts.entrySet()) {
+                        report.append("\n\t- ").append(y.getKey()).append(" (")
+                                .append(y.getValue())
+                                .append(" elements evicted due to insufficient size since last report)");
+                    }
+                    logger.warn(report.toString());
+                }
+            }
+        };
+
+        private final List<ReportTuple> entries = Collections.synchronizedList(new ArrayList<ReportTuple>());
+
+        public final void add(ReportTuple tuple) {
+            entries.add(tuple);
+        }
+
+        public final void writeReport(Logger logger) {
+
+            // Copy & Reset the entries list
+            final List<ReportTuple> list = new ArrayList<>(entries);
+            entries.clear();
+            doReportInternal(logger, list);
+
+        }
+
+        protected abstract void doReportInternal(Logger logger, List<ReportTuple> list);
+
+    }
+
+    public void generateReports() {
+        for (Reports report : Reports.values()) {
+            report.writeReport(logger);
+        }
+    }
+
+    @Async
+    public void reportCacheSizeNotLargeEnough(Ehcache cache, Element element) {
+        final ReportTuple tuple = new ReportTuple(cache, element);
+        Reports.CACHE_SIZE_NOT_LARGE_ENOUGH.add(tuple);
+    }
+
+    /*
+     * Nested Types
+     */
+
+    private static final class ReportTuple {
```

Comment:
I don't belong to the school of thought that holds there is no place for inner classes in Java.  I understand there are others -- smart folks -- who feel precisely that way... but I feel (strongly) otherwise.

I **do**, on the other hand, recognize & agree that inner classes can be (a) overused and/or (b) confusing if abused.

This inner class (`ReportTuple`) is 100% implementation detail of the surrounding class (`CacheHealthReporterService`).  It is not visible to other classes -- even in the same package -- nor is it referenced anywhere outside the class... even in configuration.  This is the kind of use case where inner classes can shine.  I would say an inner class here is an acceptable -- though not mandatory -- approach.